torch/lib/c10d/ProcessGroupGloo.cpp

```diff
@@ -363,6 +364,12 @@ void ProcessGroupGloo::SendWork::wait() {
   if (exception_) {
     std::rethrow_exception(exception_);
   }
+  return sendCompleted;
+}
+
+void ProcessGroupGloo::SendWork::abort() {
+  std::unique_lock<std::mutex> lock(mutex_);
```

Comment:
If a thread is blocked on `buffer_->waitSend()` in `wait()`, there is no way to abort since you need to hold the lock here? I'm wondering if this resolves the problem we're seeing, since I think the abort() call would just block forever and can't really interrupt `waitSend`.

Reply:
we would like to remove this lock, if

- no other thread can assign to buffer_
- abortWaitSend is thread safe to buffer_

Reply:
Synced with @xush6528  about this, we should be able to remove the locks here, `unbound_buffer` from gloo should be thread safe. We think the tests in this diff due to the way the threads were scheduled